### PR TITLE
Flatten the LocationHierarchy endpoint response

### PR DESCRIPTION
**Pull request by lincmba**
Add option to return location hierarchy in a list instead of a tree depending on a parameter value

**IMPORTANT: Where possible all PRs must be linked to a Github issue**

fixes https://github.com/opensrp/fhircore/issues/3161
fixes https://github.com/onaio/fhir-gateway-extension/issues/45

**Engineer Checklist**

- [x] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [x] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [x] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [x] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 1]
```diff
@@ -111,4 +115,43 @@ public List<Location> descendants(String locationId, Location parentLocation) {
...
+                        ? Integer.parseInt(pageSize)
+                        : Constants.PAGINATION_DEFAULT_PAGE_SIZE;
+        int page =
+                pageNumber != null
+                        ? Integer.parseInt(pageNumber)
+                        : Constants.PAGINATION_DEFAULT_PAGE_NUMBER;
+
+        int start = Math.max(0, (page - 1)) * count;
+        Location parentLocation = getLocationById(identifier);
+        List<Location> locations = getDescendants(identifier, parentLocation);
```
Comment (lincmba) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:136]: Should we consider caching this?

Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:136]: Yeah I think this is a good idea since we are caching the call to `getLocationHierarchy` for some great optimization. 

Because this is variation of the endpoint is paginated, you may want to add that(pagination) info as part of the key e.g. `<locationid>-<pageNo>`. This will probably only need to apply when the user has provided pagination

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 2]
```diff
@@ -357,16 +357,28 @@ private Bundle postProcessModeListEntries(
...
-        resultBundle.setTotal(requestBundle.getEntry().size());
+        StringBuilder urlBuilder = new StringBuilder(request.getFhirServerBase());
+        urlBuilder.append("/").append(request.getRequestPath());
 
-        // add pagination links
-        int nextPage = page < totalEntries / count ? page + 1 : 0; // 0 indicates no next page
+        return addPaginationLinks(urlBuilder, resultBundle, page, totalEntries, count, parameters);
+    }
+
+    public static Bundle addPaginationLinks(
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:366]: Since this method is being used by multiple classes let us create and move it to a generic Utils class.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/BaseEndpoint.java [hunk 3]
```diff
@@ -11,7 +11,7 @@
 import org.hl7.fhir.r4.model.Resource;
 
 public abstract class BaseEndpoint extends HttpServlet {
-    public Bundle createBundle(List<Resource> resourceList) {
+    public static Bundle createBundle(List<Resource> resourceList) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/BaseEndpoint.java:14]: Since these are now being used by multiple classes lets move them to the Utils class aforementioned

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/BaseEndpoint.java [hunk 4]
```diff
@@ -24,7 +24,7 @@ public Bundle createBundle(List<Resource> resourceList) {
         return responseBundle;
     }
 
-    public Bundle createEmptyBundle(String requestURL) {
+    public static Bundle createEmptyBundle(String requestURL) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/BaseEndpoint.java:27]: Same as above

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 5]
```diff
@@ -111,4 +115,43 @@ public List<Location> descendants(String locationId, Location parentLocation) {
...
+        List<Resource> resourceLocations = new ArrayList<>(locations);
+        int totalEntries = locations.size();
+
+        int end = Math.min(start + count, resourceLocations.size());
+        List<Resource> paginatedResourceLocations = resourceLocations.subList(start, end);
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:141]: In future we may want to figure out an alternative approach to avoid having to re fetch the whole hierarchy for every single page (and then get the sublist). For now once we add the caching it will somewhat alleviate this.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 6]
```diff
@@ -70,10 +74,10 @@ public LocationHierarchy getLocationHierarchyCore(String locationId) {
     }
 
     private List<Location> getLocationHierarchy(String locationId, Location parentLocation) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:76]: Let's rename to:
```suggestion
    private List<Location> getLocationHierarchyLocations(String locationId, Location parentLocation) {
```

**Comment by ndegwamartin** [comment]
Lets also bump up the release version so that we can publish once the PR is ready for merge.

**Comment by ndegwamartin** [comment]
@lincmba lets also bump up the release version so that we can publish once the PR is ready for merge.

**Review by ndegwamartin — APPROVED** [review]
